### DISCUSSION TITLE
Custom DbContext

### DISCUSSION THREAD

**Discussion by vinujosey** [elsa-workflows/elsa-core#5662] [Q&A]
I am trying to use custom db context for elsa as below

public class MyCustomDbContext : DbContext
{
    public MyCustomDbContext(DbContextOptions options) : base(options)
    {
    }

    // Your custom DbSets...
    public DbSet<WorkflowInstance> WorkflowInstances { get; set; }
    public DbSet<WorkflowDefinition> WorkflowDefinitions { get; set; }
    // Other Elsa-related DbSets...
}

// ...

builder.Services.AddElsa(elsa =>
{
    // Use your custom DbContext for the management layer.
    elsa.UseWorkflowManagement(management => management.UseEntityFrameworkCore<MyCustomDbContext>());

    // Use your custom DbContext for the runtime layer.
    elsa.UseWorkflowRuntime(runtime => runtime.UseEntityFrameworkCore<MyCustomDbContext>());
});

But am getting a compile error when using the extension UseEntityFrameworkCore<MyCustomDbContext>(), WorklfowManagementFeature does not contain a definition for UseEntityFrameworkCore

**Comment by glime-ai** [comment]
To resolve the compile error you're encountering when trying to use your custom `DbContext` with Elsa, you should ensure that you have the correct Elsa packages installed that support Entity Framework Core integration. Additionally, you need to make sure that your custom `DbContext` includes the necessary Elsa-related `DbSet` properties.

The `UseEntityFrameworkCore<TContext>()` extension method is used to configure Elsa to use a custom `DbContext`. However, if you're getting a compile error, it could be due to several reasons:

1. You might be missing the necessary Elsa Entity Framework Core NuGet package. Make sure you have installed the `Elsa.Persistence.EntityFramework.Core` package.

2. Your custom `DbContext` might be missing some required `DbSet` properties that Elsa expects. Ensure that your `DbContext` includes `DbSet` properties for all Elsa-related entities, such as `WorkflowInstance`, `WorkflowDefinition`, `Bookmark`, `Trigger`, etc.

3. There might be a version mismatch between the Elsa packages and the Entity Framework Core packages you are using. Ensure that all packages are compatible with each other.

4. The method signature you are using might be incorrect or outdated. Check the official Elsa documentation or the source code for the correct method signature.

Here's an example of how you might configure your custom `DbContext` with Elsa:

```csharp
public class MyCustomDbContext : DbContext
{
    public MyCustomDbContext(DbContextOptions<MyCustomDbContext> options) : base(options)
    {
    }

    // Your custom DbSets...
    public DbSet<WorkflowInstance> WorkflowInstances { get; set; }
    public DbSet<WorkflowDefinition> WorkflowDefinitions { get; set; }
    // Other Elsa-related DbSets...
}

// ...

builder.Services.AddElsa(elsa =>
{
    // Use your custom DbContext for the management layer.
    elsa.UseWorkflowManagement(management => management.UseEntityFrameworkCore<MyCustomDbContext>());

    // Use your custom DbContext for the runtime layer.
    elsa.UseWorkflowRuntime(runtime => runtime.UseEntityFrameworkCore<MyCustomDbContext>());
});
```

If you continue to experience issues, please provide the exact error message you're receiving, and ensure that your project references the correct versions of the Elsa packages.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Reply by gustavoaraujocardozo** [reply]
i'm facing the exactly same inssue.
![image](https://github.com/elsa-workflows/elsa-core/assets/11926577/ad3feed3-a606-4f0c-813d-79ffe2e35440)
it seems WorkflowManagementFeatureExtensions doesn't accept the first argument.
You can see the native line above, and my code bellow.

**Reply by rakeshkhoodeeram** [reply]
Hello,
Did you manage to make it work with MySQL?
Having the same issue with my own dbcontext.

**Reply by dangphung4** [reply]
Hi, having similar issues...

**Reply by gustavoaraujocardozo** [reply]
Scroll down to the next answer it may be helpful

**Comment by vinujosey** [comment]
I am using the version 3.1.3, does this version has the extension method UseEntityFrameworkCore<TContext>() ? Or Do I need to use an upgraded version ? When I checked the source code in git hub I couldnt find this extension method in the class Elsa.EntityFrameworkCore.Modules.Management.WorkflowManagementFeatureExtensions

**Reply by gustavoaraujocardozo** [reply]
I don't know if works to you, but for me this sample is working fine. I have set the project to run with postgres by using this sample. Check out.

https://github.com/elsa-workflows/elsa-core/blob/main/samples/aspnet/Elsa.Samples.AspNet.EntityFrameworkCore.PostgresSql/Program.cs

![image](https://github.com/elsa-workflows/elsa-core/assets/11926577/7ed8088e-580e-4470-8577-60e07d1fc97c)


Make sure to use "using Elsa.EntityFrameworkCore.Common;" and the package Elsa.EntityFrameworkCore.PostgreSql

**Comment by sfmskywalker** [comment]
`UseEntityFrameworkCore<TContext>()` is not an Elsa 3.1.3 API. The EF Core extensions accept an optional configuration callback; they do not accept an application-defined `DbContext` type. So the compile error is expected.

Use Elsa's management and runtime EF Core contexts and configure the provider on each feature, for example:

```csharp
builder.Services.AddElsa(elsa => elsa
    .UseWorkflowManagement(management =>
        management.UseEntityFrameworkCore(ef => ef.UseSqlServer(connectionString)))
    .UseWorkflowRuntime(runtime =>
        runtime.UseEntityFrameworkCore(ef => ef.UseSqlServer(connectionString))));
```

Make sure the matching Elsa `EntityFrameworkCore` provider package and its extension namespaces are referenced. Sharing one custom `DbContext` with Elsa's management/runtime stores is not supported by that generic overload, because no such overload exists.

The 3.1.3 implementation is in `src/modules/Elsa.EntityFrameworkCore/Modules/Management/Extensions.cs` (and the runtime extension alongside it); its `UseEntityFrameworkCore` signatures take configuration delegates only.
